addon/mixins/ember-hooks.ts

```diff
@@ -0,0 +1,228 @@
+    return obj;
+  }
+
+  return obj;
+}
+
+const observable = (obj, scope, ancestors) => {
+  if (typeof obj === 'object') {
+    // Set metadata
+    Object.defineProperty(obj, '__isProxy', {
+      value: true,
+      enumerable: false,
+    });
+    Object.defineProperty(obj, '__ancestors', {
+      value: ancestors,
+      enumerable: false,
+    })
+
+    Object.keys(obj).forEach(key => {
+      const prev = obj.__ancestors ? [...obj.__ancestors, key] : [key];
+      obj[key] = observable(obj[key], scope, prev);
+    });
+
+    return new Proxy(obj, {
+      get(target, prop) {
+        if (target.__isProxy) {
+          if (Array.isArray(target)) {
+            return _handleArrayFunctions(scope, target, prop);
+          }
+          return target[prop];
+        }
+
+        return scope.get(prop);
+      },
+      set(obj, prop, value) {
+        if (obj.__ancestors) {
+          const navigationString = createPropertyNavigationString(obj.__ancestors, prop);
+          scope.set(navigationString, value);
+        } else {
+          scope.set(prop, value);
+        }
+
+        // TODO: tw - Reflect?
+        obj[prop] = value;
+
+        // Set should return true if it was successful
+        return true;
+      },
+    });
+  } else {
+    return obj;
+  }
+};
+
+const _handleArrayFunctions = (scope, target, prop) => {
+  // TODO: tw - should this be an object instead of ifs?
+  const emberArray = scope.get(target.__ancestors.join());
+  if (prop === 'push') {
+    return (val) => {
+      emberArray.pushObject(val);
+      return target.push(val);
+    };
+  } else if (prop === 'pop') {
+    return () => {
+      emberArray.popObject();
+      return target.pop();
+    }
+  } else if (prop === 'shift') {
+    return () => {
+      emberArray.shiftObject();
+      return target.shift();
+    }
+  } else if (prop === 'unshift') {
+    return (...args) => {
+      emberArray.unshiftObjects(args);
+      return target.unshift(...args);
+    }
+  } else if (prop === 'reverse') {
+    return () => {
+      emberArray.reverseObjects();
+      return target.reverse();
+    }
+  }
+
+  // TODO: tw- should this have a default return, if so what?
+}
+
+const createPropertyNavigationString = (ancestors, prop) => `${ancestors.join('.')}.${prop}`;
+
+export const useStore = () => {
+  const self = currentInstance;
+  if (self._state === 'preRender') self.set('store', service('store'));
+  return self.get('store');
+};
+
+/**
+ * useMemo is a hook that will only recompute a potentially expensive
+ * value when one of it's dependencies change.
+ *
+ * @param {function} createMemoizedValue - function to compute an expensive value
+ * @param {Array} deps - Array of dependencies
+ */
+export const useMemo = <T>(createMemoizedValue: () => T, deps?: any[] | null): T => {
+  const instance = currentInstance;
+  if (instance === null) {
+    throw new Error('Unable to find Ember instance');
+  }
+
+  hookCallIndex++;
+  deps = deps != null ? deps : null;
```

Comment:
Why is this line necessary?

Reply:
I believe it has to do w/ typescript validation. @jcampuza ?

Reply:
That’s what I thought at first but the type after this statement will still be <any[] | null>. When th conditional down below checks for null, that’s when typescript will recognize it as <any[]>

Reply:
It's to force deps to be of type `<any[] | null>`. Before this statement it would be of type `<any[] | null | undefined>`. In essence it's casting it so that it's not undefined.

Reply:
React's hooks implementation does a similary check, a lot of this was inspired by that.

Reply:
Thanks! I missed the `?` in the input params.